File: using_C++/64_minpath.cpp
using namespace std;
static const auto _ = []()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    return nullptr;
}();
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {

        int m = grid.size();
        int n = grid[0].size();
        for (int i = 0;i<m;i++){
            for (int j =0; j<n;j++){
                if (i ==0 && j==0) continue;
                else if (i==0 && j>0){
                    grid[i][j] = grid[i][j-1]+grid[i][j];
                }
                else if (i>0 && j==0){
                    grid[i][j] = grid[i][j]+grid[i-1][j];
                }

                else if (i>0 && j>0){
                    grid[i][j] = min(grid[i-1][j],grid[i][j-1])+grid[i][j];
                }
            }
        }
        return grid[m-1][n-1];
    }
};

File: using_python/64_minpath.py
# 输入:
# [
#   [1,3,1],
#   [1,5,1],
#   [4,2,1]
# ]
# 输出: 7

class Solution:
    def minPathSum(self, grid: [[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        dp = [[65535] * n for _ in range(m)]
        dp[0][0] = grid[0][0]
        ### dp[i][j] = min(dp[i][j-1],dp[i-1][j])
        for i in range(m):
            for j in range(n):
                if i == 0 and j == 0:
                    continue
                dp[i][j] = grid[i][j] + min(dp[i][j - 1], dp[i - 1][j])

        return dp[-1][-1]


test = [
    [1, 3, 1],
    [1, 5, 1],
    [4, 2, 1]
]
so = Solution()
print(so.minPathSum(test))
